smaily_for_drupal/src/Form/AdminForm.php

```diff
@@ -86,13 +86,41 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       ],
     ];
 
+    $form['smaily_container']['smaily_custom_fields'] = [
+      '#type' => 'fieldset',
+      '#description' => $this->t('List of key-value pairs on separated rows,
+        that will be displayed on the registration form as checkboxes.
+        ie. subscription1|Main news'),
+      '#title' => $this->t('Custom fields'),
+    ];
+
+    $form['smaily_container']['smaily_custom_fields']['customfields'] = [
+      '#type' => 'textarea',
+      '#default_value' => $config->get('smaily_custom_fields', ''),
+    ];
+
     $form['smaily_container']['submit'] = [
       '#type' => 'submit',
       '#value' => $this->t('Save settings'),
     ];
     return $form;
   }
 
+  /**
+   * {@inheritdoc}
+   */
+  public function validateForm(array &$form, FormStateInterface $form_state) {
+    $lines = $form_state->getValue('customfields');
+    if (!empty($lines)) {
+      foreach (explode("\n", $lines) as $line) {
+        // Match any non-whitespace character with "|" and with any non-whitespace character.
+        if (!preg_match('/[\S+]\|[\S+]/', $line)) {
```

Comment:
This regular expression will allow value `foo|bar|Baz` through. Where field name will be `foo|bar` and value `Baz`, but it shouldn't be allowed to save at all. Also user might have this kind of value there also - `foo|My favourite character is pipe (|)` - safest option would be to not allow user to save this value.

Reply:
Updated the pattern. Is it sufficient?

Reply:
Yep.